### DISCUSSION TITLE
Do you think it is good to avoid useCallback with zustand

### DISCUSSION THREAD

**Discussion by MetinArslanturk** [pmndrs/zustand#1061] [Q&A] [answered]
Hi,
I mean lets' assume that we have `id`, `setUser` and `user` fields in zustand store. We have a function to fetch user with id. We will trigger this function in event handler of a child component (ie onClick), so normally we would want to wrap it with useCallback.  (Taking ID as function argument would be nicer I know, but leaving like this to make example more correct)

```
const MyComponent = () => {

   const id = useStore(state => state.id);
   const setUser = useStore(state => state.setUser;
   
   const fetchUser = useCallback(
      () => {
        const user = await fetch('http..../' + id);
        setUser(user);
      },
      [id, setUser],
    )

// .....
}
```

In above, normally setUser function's reference will never change but id will change. So, when every id variable change (re-render) we will generate a new function with useCallback.

But, the good part of zustand is that we can reach the store outside of react. So what if we completely move this function outside of component and getState snapshot when we call function.

```
const fetchUser = () => {
   const {id, setUser} = useStore.getState()
   const user = await fetch('http..../' + id);
    setUser(user);
 }

const MyComponent = () => {
// .....
}
```

So, I avoided useCallback and regeneration of my function. While playing with data in the store for an event handler, this approach seemed me to nicer. But I am not sure this has any disadvantage or risk. What are your opinions I wanted to ask.

**Comment by dai-shi** [comment, marked as answer]
No risks.
Historically, we educate the hook style (your first example) and avoid using `.getState()` and `.setState()` apis.
But, your second example is more performant and if you like the style, no disadvantages.

Slightly related, I would like to educate another approach: https://github.com/pmndrs/zustand/wiki/Practice-with-no-store-actions
